### DISCUSSION TITLE
Feature Request: Support Insert or Ignore when doing bulk inserts

### DISCUSSION THREAD

**Discussion by bjornasm** [duckdb/duckdb#8817] [Ideas]
Insert or ignore seems to be supported for simple statements:

```sql 
CREATE TABLE test (a integer, primary key(a))
INSERT INTO test values (1);
INSERT INTO test values (2);
INSERT INTO test values (1);
Error: Constraint Error: Duplicate key "a: 1" violates primary key constraint. If this is an unexpected constraint violation please double check with the known index limitations section in our documentation (docs - sql - indexes).
INSERT OR IGNORE INTO test values(2);
```

However, when doing bulk inserts (here exemplified with a similar table that has duplicate values):
```sql
CREATE TABLE test (a integer, primary key(a));
CREATE TABLE test2(b integer);
INSERT INTO test2 values(2);
INSERT INTO test2 values(2);
INSERT OR IGNORE INTO test SELECT * FROM test2;
Error: Constraint Error: Duplicate key "a: 2" violates primary key constraint. If this is an unexpected constraint violation please double check with the known index limitations section in our documentation (docs - sql - indexes).
```

So - when bulk inserting from files or other tables you do not have insert or ignore functionality. An implementation for the users would be to check for duplicates themselves, f.ex by performing an anti-join in addition to ensure that the values you are inserting is not already present, then distinct or groupby to ensure that you are not attempting to insert two new but duplicate rows: using `distinct` if all columns are the same as the primary keys, or `group by` if you have several columns in addition to the ones that are primary keys. 

```sql
 INSERT INTO test SELECT distinct b FROM test2 ANTI JOIN test ON test.a = test2.b;
````
or
```sql
 INSERT INTO test SELECT b, any_value(c), any_value(d) FROM test2 ANTI JOIN test ON test.a = test2.b group by b;
````
edit: (Or just a DISTINCT ON (primarykey1, primarykey2) in addition to an anti join, thx @Tishj )

However this requires that you know the primary keys of the table you are inserting into. While the functionality could be the same, could this be implemented in DuckDB so DuckDB performs the anti-join (using the information schema to figure out the primary keys) and thus accept using INSERT OR IGNORE? Admittedly, a caveat against implementing it this way is that there is a choice to be made when making the set you are inserting unique (any_value(), mode() etc) - so that might be something that the user should have to decide themselves.

**Comment by Tishj** [comment]
I dont follow, you state this as the issue
```
CREATE TABLE test (a integer, primary key(a))
CREATE TABLE test2(b integer)
INSERT INTO test2 values(2);
INSERT INTO test2 values(2);
INSERT INTO test SELECT * FROM test2;
```

But this doesnt contain INSERT OR IGNORE?

**Reply by bjornasm** [reply]
Sorry, I miswrote my statement when I created this post - updated my original post to INSERT OR IGNORE:

```D CREATE TABLE test (a integer, primary key(a));
Run Time (s): real 0.001 user 0.000339 sys 0.000021
D CREATE TABLE test2(b integer);
Run Time (s): real 0.001 user 0.000102 sys 0.000007
D INSERT INTO test2 values(2);
Run Time (s): real 0.001 user 0.000461 sys 0.000027
D INSERT INTO test2 values(2);
Run Time (s): real 0.000 user 0.000500 sys 0.000030
D INSERT INTO test SELECT * FROM test2;
Run Time (s): real 0.002 user 0.000730 sys 0.000044
Error: Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicate key "2"
D INSERT OR IGNORE INTO test SELECT * FROM test2;
Run Time (s): real 0.001 user 0.000477 sys 0.000029
Error: Constraint Error: PRIMARY KEY or UNIQUE constraint violated: duplicate key "2"
```

**Reply by Tishj** [reply]
Ah, yes now I understsnd

INSERT OR IGNORE can resolve conflicts between existing data and to-be-inserted data

Your conflicts are within the to-be-inserted data, currently this is expected behavior and you should add a DISTINCT to your selection

**Reply by bjornasm** [reply]
Thanks! You are partially right, however a DISTINCT may not help, in example:

```
| A | B | C
 1    2   4
 1    2   5

```
Both of these rows are distinct, however if your primary key is a combination of column A and B this will throw a constraint error.

**Reply by Tishj** [reply]
Correct, then you would need DISTINCT ON (a, b)

**Reply by bjornasm** [reply]
Ah, good call!

**Comment by rlancer** [comment]
I would expect `ignore` or `replace` to behave the same regardless of if the data is bulk added or not. In the minimum the current behavior should be documented
